### PR TITLE
Adding a new lab for text classification using BERT 

### PR DESCRIPTION
**Pull request by sanjanalreddy**
Adding a new lab and solution based on the tutorial to do sentiment analysis on IMDB dataset using small BERT

### CODE REVIEW

File: courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb [hunk 1]
```diff
@@ -0,0 +1,1190 @@
+{
```
Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>Nit: add Learning Objectives heading for consistency with ASL labs above "in this notebook" line.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>I'm torn between the flexibility this lab offers and the potentially overwhelming number of BERT and related model variants available. Since ASL labs are generally targeting 45 min - 1 hour, maybe consider being more perspective? Some options here are to to reduce set below and keep the links you have here. Others might be just setting up the small bert example first and then having a continue your learning section below that has the rest of the models for students to experiment with. Shop this around with the team for alternative opinions for sure.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>Nit: Maybe consider breaking up this print block and mixing it with a one sentence explanation of each? E.g. I was unclear what the type ID was on first pass from the output thinking something was wrong before reading 4 lines down what it referred to.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>Clarify ordering of layers in model. Some other helpful hints to students might be reiterating you will be using the pooled output accessible using the dictionary syntax like above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>Nit: this could be potentially reworded for further clarity. I read this like the following but could just be me:</p><p> </p><p>"For fine-tuning, let's use the same optimizer that BERT was originally trained with: the "Adaptive Moments" (Adam)." -&gt; ok, so Adam? Thought we imported AdamW above?</p><p> </p><p>This optimizer minimizes the prediction loss and does regularization by weight decay (not using moments), which is also known as <a href="https://arxiv.org/abs/1711.05101" target="_blank">AdamW</a>. -&gt; Ah no ok it was AdamW, not just moments for updates but uses some kind of weight decay so actually AdamW variant.</p><p> </p><p>Since Adam has kinda just been the default across labs, maybe a sentence somewhere mentioning AdamW is a newer optimizer meant to improve generalization</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

File: courses/machine_learning/deepdive2/text_classification/solutions/classify_text_with_bert.ipynb [hunk 2]
```diff
@@ -0,0 +1,1392 @@
+{
```
Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/solutions/classify_text_with_bert.ipynb:1]: <p>Consider ordering the text layers to match the layers in the model e.g. switch dense and dropout</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/solutions/classify_text_with_bert.ipynb:1]: <p>activation=none defaults to linear right? Since you apply the sigmoid below to normalize the positive sentiment class probability, what do you think about moving up to condense? Or making that explicit activation=linear for students to make that connection</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (dougkelly) [inline courses/machine_learning/deepdive2/text_classification/solutions/classify_text_with_bert.ipynb:1]: <p>Looks like a bit of overfitting. Maybe increase the dropout a bit and add earlystopping callback to help with that and possibly even speed up the lab runtime a bit?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1412/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

**Comment by dougkelly** [comment]
Thanks Sanjana! This is a great lab addition. Few nits on styling and suggestions for further clarity. Happy to talk over any of the suggestions; feel free to make the changes you would like and merge ahead of session.

File: courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb (hunk 1 above)
Comment (sanjanalreddy) [reply courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>Thanks for this suggestion! I decided to add a continued learning section at the end for the curious students for now.</p>

Comment (sanjanalreddy) [reply courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>I added a list and clarified the pooled output. Added it as a hint. This was a really good suggestion since this is the core part of the lab.</p>

Comment (sanjanalreddy) [reply courses/machine_learning/deepdive2/text_classification/labs/classify_text_with_bert.ipynb:1]: <p>Added clarification and also a link to the tensorflow documentation showing how it is implemented.</p>

File: courses/machine_learning/deepdive2/text_classification/solutions/classify_text_with_bert.ipynb (hunk 2 above)
Comment (sanjanalreddy) [reply courses/machine_learning/deepdive2/text_classification/solutions/classify_text_with_bert.ipynb:1]: <p>That's right, it would be a linear layer as the last layer. I moved the code cell with Sigmoid up and then added activation=linear as you suggested since that is what they have seen through ASL.</p>

**Comment by dougkelly** [comment]
LGTM. Merging.
